rllib/utils/nested_dict.py

```diff
@@ -0,0 +1,280 @@
+        that a nested dict can take.
+
+    Args:
+        x: a representation of a nested dict: it can be an iterable of `StreamType`
+        to values. e.g. `[(("a", "b") , 1), ("b", 2)]` or a mapping of flattened
+        keys to values. e.g. `{("a", "b"): 1, ("b",): 2}` or any nested mapping,
+        e.g. `{"a": {"b": 1}, "b": 2}`.
+
+    Example:
+        Basic usage:
+            >>> foo_dict = NestedDict()
+            >>> # Setting elements, possibly nested:
+            >>> foo_dict['a'] = 100         # foo_dict = {'a': 100}
+            >>> foo_dict['b', 'c'] = 200    # foo_dict = {'a': 100, 'b': {'c': 200}}
+            >>> foo_dict['b', 'd'] = 300    # foo_dict = {'a': 100,
+            >>>                             #             'b': {'c': 200, 'd': 300}}
+            >>> # Getting elements, possibly nested:
+            >>> print(foo_dict['b', 'c'])   # 200
+            >>> print(foo_dict['b']) # IndexError("Use get for partial indexing.")
+            >>> print(foo_dict.get('b'))    # {'c': 200, 'd': 300}
+            >>> print(foo_dict) # {'a': 100, 'b': {'c': 200, 'd': 300}}
+            >>> # Converting to a dict:
+            >>> foo_dict.asdict()  # {'a': 100, 'b': {'c': 200, 'd': 300}}
+            >>> # len function:
+            >>> print(len(foo_dict))  # 3
+            >>> # Iterating:
+            >>> foo_dict.keys()  # dict_keys(['a', ('b', 'c'), ('b', 'd')])
+            >>> foo_dict.items() # dict_items([('a', 100), (('b', 'c'), 200), (('b',
+            'd'), 300)])
+            >>> foo_dict.shallow_keys()  # dict_keys(['a', 'b'])
+        Filter:
+            >>> dict1 = NestedDict([
+                (('foo', 'a'), 10), (('foo', 'b'), 11),
+                (('bar', 'c'), 11), (('bar', 'a'), 110)])
+            >>> dict2 = NestedDict([('foo', NestedDict(dict(a=11)))])
+            >>> dict3 = NestedDict([('foo', NestedDict(dict(a=100))),
+                                    ('bar', NestedDict(dict(d=11)))])
+            >>> dict4 = NestedDict([('foo', NestedDict(dict(a=100))),
+                                    ('bar', NestedDict(dict(c=11)))])
+            >>> dict1.filter(dict2).asdict()   # {'foo': {'a': 10}}
+            >>> dict1.filter(dict4).asdict()   # {'bar': {'c': 11}, 'foo': {'a': 10}}
+            >>> dict1.filter(dict3).asdict()   # KeyError - ('bar', 'd') not in dict1
+    """
+
+    def __init__(
+        self,
+        x: Optional[NestedDictInputType] = None,
+    ):
+        self._data = dict()  # type: Dict[str, Any]
+        x = x or {}
+        if isinstance(x, Mapping):
+            for k, v in x.items():
+                self[k] = v
+        else:
+            if not isinstance(x, Iterable):
+                raise ValueError(f"Input must be a Mapping or Iterable, got {x}.")
+            for k, v in x:
+                self[k] = v
+
+    def __contains__(self, k: _IndexType) -> bool:
+        k = _flatten_index(k)
```

Comment:
we could increase space complexity and store the flattened index every single time a key is updated.

Reply:
good call. Took me a lot of thinking and optimization to convert everything to iterative instead of recursive. updated all the unittests to be more comprehensive.